csharp/src/Ice/OutputStream.cs

```diff
@@ -1009,95 +662,390 @@ public void WriteClass(int tag, AnyClass? v)
             }
         }
 
-        public void WriteSeq<T>(T[] v, OutputStreamStructWriter<T> writer) where T : struct
+        /// <summary>Writes a tagged variable-size struct to the stream.</summary>
+        /// <param name="tag">The tag.</param>
+        /// <param name="v">The struct to write.</param>
+        public void WriteTaggedStruct<T>(int tag, T? v) where T : struct, IStreamableStruct
         {
-            WriteSize(v.Length);
-            foreach (T item in v)
+            if (v is T value && WriteOptional(tag, OptionalFormat.FSize))
             {
-                writer(this, item);
+                var pos = StartSize();
+                value.IceWrite(this);
+                EndSize(pos);
             }
         }
 
-        public void WriteSeq<T>(IReadOnlyCollection<T> v, OutputStreamWriter<T> writer)
+        /// <summary>Writes a sequence to the stream.</summary>
+        /// <param name="v">The sequence to write.</param>
+        /// <param name="writer">The delegate that writes each element to the stream.</param>
+        public void WriteSeq<T>(IEnumerable<T> v, OutputStreamWriter<T> writer)
         {
-            WriteSize(v.Count);
+            WriteSize(v.Count()); // potentially slow Linq Count()
             foreach (T item in v)
             {
                 writer(this, item);
             }
         }
 
-        public void WriteSeq<T>(IReadOnlyCollection<T> v, OutputStreamStructWriter<T> writer) where T : struct
+        /// <summary>Writes a sequence to the stream. Elements of the sequence are mapped Slice structs.</summary>
+        /// <param name="v">The sequence to write.</param>
+        /// <param name="writer">The delegate that writes each element to the stream.</param>
+        public void WriteSeq<T>(IEnumerable<T> v, OutputStreamStructWriter<T> writer)
+            where T : struct, IStreamableStruct
         {
-            WriteSize(v.Count);
+            WriteSize(v.Count()); // potentially slow Linq Count()
             foreach (T item in v)
             {
                 writer(this, item);
             }
         }
 
-        public void WriteDict<TKey, TValue>(IReadOnlyDictionary<TKey, TValue> dict, OutputStreamWriter<TKey> keyWriter,
+        /// <summary>Writes a tagged sequence of fixed-size elements to the stream.</summary>
+        /// <param name="tag">The tag.</param>
+        /// <param name="v">The sequence to write.</param>
+        /// <param name="elementSize">The fixed size of each element of the sequence, in bytes.</param>
+        /// <param name="writer">The delegate that writes each element to the stream.</param>
+        public void WriteTaggedSeq<T>(int tag, IEnumerable<T>? v, int elementSize, OutputStreamWriter<T> writer)
+        {
+            Debug.Assert(elementSize > 0);
+            if (v is IEnumerable<T> value && WriteOptional(tag, OptionalFormat.VSize))
+            {
+                int count = v.Count(); // potentially slow Linq Count()
```

Comment:
should this be `value.Count()`? maybe we should avoid introducing new variables for reference types.

Reply:
Yes, fixing.